### PR TITLE
New suggest peer by address gap SWIP

### PR DESCRIPTION
**Pull request by kortatu**
Peer suggestion inside a bin is been done randomly (or the first in line). Instead of doing that, it is better to suggest a peer that covers a bigger address gap space.

### CODE REVIEW

File: SWIPs/swip-suggest-by-address-gap.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+Peer suggestion inside a bin is been done randomly (or the first in line). Instead of doing that, it is better to suggest
+a peer that covers a bigger address gap space.
+
+## Abstract
+This PR propose an improved method for suggesting peers for connecting. Whenever a peer of a given bin needs to be added,
+instead of taking the first `callable`, the method suggest the peer which covers a bigger address space gap. 
+An address space gap is defined as a set of addresses in which the current node doesn't have a connected peer.
+
+## Motivation
+Improve the suggested peer algorithm, trying to covering a bigger part of the address space and allowing the node to be
```
Comment (aperonam) [inline SWIPs/swip-suggest-by-address-gap.md]:
```suggestion
Improve the suggested peer algorithm, trying to cover a bigger part of the address space and allowing the node to be 
```

File: SWIPs/swip-suggest-by-address-gap.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+a peer that covers a bigger address gap space.
+
+## Abstract
+This PR propose an improved method for suggesting peers for connecting. Whenever a peer of a given bin needs to be added,
+instead of taking the first `callable`, the method suggest the peer which covers a bigger address space gap. 
+An address space gap is defined as a set of addresses in which the current node doesn't have a connected peer.
+
+## Motivation
+Improve the suggested peer algorithm, trying to covering a bigger part of the address space and allowing the node to be 
+more efficient when searching nearest peer to a certain address.
```
Comment (aperonam) [inline SWIPs/swip-suggest-by-address-gap.md]:
```suggestion
more efficient when searching for the nearest peer to a certain address.
```

File: SWIPs/swip-suggest-by-address-gap.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+## Rationale
+In order to increment performance of content retrieval and delivery the node should minimize the size of its gaps, because this 
+means that it knows peers near almost all addresses. If the minimum `gap` in the kademlia table is 4, it means that whatever 
+look up or forwarding done will be at least 4 po far away. On the other hand, if the node has a 0 po `gap`, it means that
+for half the addresses, the next jump will be still 0 po away!.
+
+
+### Gaps for peer suggestion
+The current kademlia bootstrap algorithm try to fill in the bins (or po spaces) until some level of saturation is reached.
```
Comment (aperonam) [inline SWIPs/swip-suggest-by-address-gap.md]:
```suggestion
The current kademlia bootstrap algorithm tries to fill in the bins (or po spaces) until some level of saturation is reached.
```

File: SWIPs/swip-suggest-by-address-gap.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+In the process of doing that, the `gaps` will diminish, but not in the optimal way.
+
+For example, if the node address is `00000000`, it is connected only with one peer in bin 0 `10000000` and the known 
+addresses for bin 0 are: `10000001` and `11000000`. The current algorithm we will take the first `callable` one, so
```
Comment (aperonam) [inline SWIPs/swip-suggest-by-address-gap.md]:
```suggestion
addresses for bin 0 are: `10000001` and `11000000`. The current algorithm will take the first `callable` one, so 
```

File: SWIPs/swip-suggest-by-address-gap.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+ peers serving that gap. If we have `n` current peers that are equidistant to a particular gap of size `s`,
+the load of each of these peers will be on average `s/n`. 
+We can define a gap's `temperature` as that number `s/n`. When looking for new peers to connect, instead of looking for
+bigger gaps we could look for `hotter` gaps.
+For example, if in our first example, we can't find a peer in `11xxxxxx` and we instead, used the best peer, we could end
+with the configuration in Fig. 3.
+![Fig. 3](https://raw.githubusercontent.com/kortatu/swarm_doc/master/address_space_gaps-lb-3.png)
+Fig. 3 - Comparing gaps temperature
+
+Here we still have `11xxxxxx` as the biggest gap (po=1, size 1/4), same size as `01xxxxxx`. But if consider temperature,
```
Comment (aperonam) [inline SWIPs/swip-suggest-by-address-gap.md]:
```suggestion
Here we still have `11xxxxxx` as the biggest gap (po=1, size 1/4), same size as `01xxxxxx`. But if we consider temperature,
```

File: SWIPs/swip-suggest-by-address-gap.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+`01xxxxxx` is hotter because is served only by our node `00000000`, being its temperature is `(1/4)/ 1 = 1/4`. However,
```
Comment (aperonam) [inline SWIPs/swip-suggest-by-address-gap.md]:
```suggestion
`01xxxxxx` is hotter because is served only by our node `00000000`, being its temperature  `(1/4)/ 1 = 1/4`. However,
```

File: SWIPs/swip-suggest-by-address-gap.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+`11xxxxxx` is now served by two peers, so its temperature is `(1/4) / 2 = 1/8`, and that will mean that we will select
+`01xxxxxx` as the hotter one.
+
+There is a way of implementing temperature calculation so its cost it is the same as looking for biggest gap. Temperature
```
Comment (aperonam) [inline SWIPs/swip-suggest-by-address-gap.md]:
```suggestion
There is a way of implementing temperature calculation so its cost  is the same as looking for biggest gap. Temperature
```

File: SWIPs/swip-suggest-by-address-gap.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+A preliminary implementation is done in [PR #1869](https://github.com/ethersphere/swarm/pull/1869) 
+The search for gaps can be done easily using a proximity order tree or `pot`. Traversing the bins of a node, a `gap` is
+found if there is some of the po's missing starting from furthest (left). In each level the starting po to search for is the
+parent po (not 0, because in the second level, under a node of po=0, the minimum po that could be found is 1).
+
+Implementation of the function that looks for the bigger Gap in a `pot` can be seen in
+`pot.BiggestAddressGap`. That function returns the biggest gap in the form of a po and
+a node under the gap can be found.
+
+This function is used in `kademlia.suggestPeerInBinByGap`, which it returns a BzzAddress in a particular bin which fills
```
Comment (aperonam) [inline SWIPs/swip-suggest-by-address-gap.md]:
```suggestion
This function is used in `kademlia.suggestPeerInBinByGap`, which returns a BzzAddress in a particular bin which fills
```

**Comment by Eknir** [comment]
Thanks, that was very enlighting to read. 

This SWIP is very relevant with respect to the proposed price discovery mechanism of SWAP. 
TLDR: the price of a `retrieveRequest` will depend on the number of hops it takes to retrieve this file. Any request sent to a peer that connects to other peers based on `gap` is expected to take fewer hops to complete than a peer that connects without this strategy. 

What is the status of this SWIP and it's PR?

**Comment by Eknir** [comment]
Could you shed some light for me on what the expected amount of hops is when retrieving a file for a given any log2Distance (D) between the node and the chunk; with your algorithm for filling the bins and the old one?

From what I get, parameters are the networkSize (N) and binSize (B). 

If you don't have this readily available, I can help with it as well.

**Comment by acud** [comment]
@Eknir the number of hops to retrieve a chunk does not change as a result of this SWIP, it is a property of the network which is guaranteed by kademlia routing and the network topology. This SWIP has to do more with connectivity pattern and load balancing rather than a change to the underlying premises of the network
